polymesh-docs/primitives/smart-contracts.mdx

```diff
@@ -0,0 +1,347 @@
+
+One concern with (possibly immutable) smart contracts calling into the Polymesh native layer is that Polymesh allows this native functionality (logic and storage) to be upgraded via chain upgrades. Typically these upgrades are backwards compatable, but in the case of major releases may not be.
+
+To help mitigate this Polymesh provides a proxy contract (Upgradable Polymesh Ink Contract) through which interactions with the native layer can be mediated. This provides a stable interface for smart contracts to interact with, across any breaking Polymesh upgrades.
+
+The Upgradable Polymesh Ink Contract can be found at:
+https://github.com/PolymeshAssociation/Polymesh/tree/develop/contracts/upgradeable-polymesh-ink
+and is a normal Polymesh smart contract, that has an API designed to be called by other smart contracts (rather than directly through the `Contracts::call` extrinsic).
+
+This contract is maintained by the Polymesh Association, with the intention that if there are breaking changes during an upgrade that mean any of its functionality no longer works, a new fixed version will be deployed and made available ahead of these breaking changes.
+
+It presents a clean API such as:
+```
+let api = PolymeshInk::new()?;
+// Accept portfolio custody and ensure we have custody.
+let portfolio_id = api.accept_portfolio_custody(auth_id, portfolio)?;
+```
+
+Under the covers, the call to `api.accept_portfolio_custody` uses a delegate call (`ink::env::call::build_call::<ink::env::DefaultEnvironment>().delegate`) into the Upgradable Polymesh Ink Contract, to submit (in this case) an extrinsic to accept portfolio custody.
+
+The Upgradable Polymesh Ink Contract uses the Polymesh API extensively to interact with the native Polymesh layer via storage queries and executing extrinsics.
+
+#### Updating the Upgradable Polymesh Ink Contract
+
+Polymesh provides dedicated Chain Extensions to allow smart contracts to discover the latest version of the Upgradable Polymesh Ink Contract. These Chain Extensions are used by the API associated with the Upgradable Polymesh Ink Contract to help ensure it is always using the correct underlying logic via the latest contract hash / code.
+
+In particular, from the `polymesh-ink` crate, using `let api = PolymeshInk::new()?;` will look up the latest Upgradable Polymesh Ink Contract hash from chain storage (`polymeshContracts::currentApiHash`) and use this address when dispatching calls to the Upgradable Polymesh Ink Contract via its API.
+
+Since this latest Upgradable Polymesh Ink contract hash may change (for example if a Polymesh upgrade necessitates releasing a new version with updated logic, maintaining the same external API) it is important to either re-instantiate the Polymesh Ink API everytime it is used (when constructed the API checks it is referencing the latest Upgradable Polymesh Ink Contract hash), or use the API function `check_for_upgrade` which will do the same on an already instantiated API object.
```

Comment:
User's shouldn't use the `check_for_upgrade`.  That would only be needed if they store the `PolymeshInk` value in the contract's storage, which isn't recommended anymore.  That was more important when we used a tracker contract, since the upgrade check was more expensive.  For a single call into the contract they should try to re-use the same `api` value as much as possible (at the very least don't create it inside a loop).